File: apps/api/app/schemas/etalons.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.enums import CheckStatus, DocumentType, EtalonSource, EtalonStatus, Severity, Verdict


class EtalonEvidence(BaseModel):
    quote: str = Field(min_length=1)
    location: str = Field(min_length=1)


class EtalonLayer1Item(BaseModel):
    id: str = Field(min_length=1)
    dimension: str = Field(min_length=1)
    status: CheckStatus
    severity: Severity
    title: str = Field(min_length=1)
    summary: str = Field(min_length=1)
    evidence: list[EtalonEvidence] = Field(default_factory=list)
    recommendation: str = ""
    confidence: float | None = Field(default=None, ge=0, le=1)


class EtalonLayer2Item(BaseModel):
    id: str = Field(min_length=1)
    parent_layer_1_id: str = Field(min_length=1)
    check: str = Field(min_length=1)
    status: CheckStatus
    severity: Severity
    finding: str = Field(min_length=1)
    evidence: list[EtalonEvidence] = Field(default_factory=list)
    expected_fix: str = ""
    confidence: float | None = Field(default=None, ge=0, le=1)


class EtalonPayload(BaseModel):
    expected_verdict: Verdict
    layer_1: list[EtalonLayer1Item] = Field(default_factory=list)
    layer_2: list[EtalonLayer2Item] = Field(default_factory=list)
    key_findings: list[str] = Field(default_factory=list)
    forbidden_false_findings: list[str] = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_layer_2_parents(self) -> "EtalonPayload":
        layer_1_ids = {item.id for item in self.layer_1}
        orphan_ids = sorted({item.parent_layer_1_id for item in self.layer_2 if item.parent_layer_1_id not in layer_1_ids})
        if orphan_ids:
            raise ValueError(f"Layer 2 parent_layer_1_id values do not exist in Layer 1: {', '.join(orphan_ids)}")
        return self


class EtalonDraftCreate(BaseModel):
    status: EtalonStatus = EtalonStatus.DRAFT


class Gate2BenchmarkImportRequest(BaseModel):
    benchmark_dir: str | None = None
    activate: bool = False


class EtalonUpdate(BaseModel):
    expected_verdict: Verdict | None = None
    layer_1: list[EtalonLayer1Item] | None = None
    layer_2: list[EtalonLayer2Item] | None = None
    key_findings: list[str] | None = None
    forbidden_false_findings: list[str] | None = None
    real_defense_status: str | None = None
    defense_comments: str | None = None
    raw_file_visible_to_all: bool | None = None


class EtalonRead(BaseModel):
    id: UUID
    document_id: UUID
    author_id: UUID
    source: EtalonSource
    document_type: DocumentType
    real_defense_status: str | None
    defense_comments: str | None
    expected_verdict: Verdict
    layer_1: list[EtalonLayer1Item]
    layer_2: list[EtalonLayer2Item]
    key_findings: list[str]
    forbidden_false_findings: list[str]
    source_metadata: dict
    status: EtalonStatus
    version: int
    raw_file_visible_to_all: bool
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class EtalonsListResponse(BaseModel):
    etalons: list[EtalonRead]


class Gate2BenchmarkImportResponse(BaseModel):
    imported_count: int
    skipped_count: int
    updated_count: int = 0
    unmatched_count: int = 0
    parse_enqueued_count: int
    etalons: list[EtalonRead]

File: apps/web/src/lib/api/etalons.ts
import { apiFetch } from "./client";
import type { DocumentType } from "./documents";

export type EtalonStatus = "draft" | "active" | "archived" | "deleted";
export type Verdict = "approve" | "approve_with_conditions" | "need_evidence" | "reject" | "unknown";
export type CheckStatus = "pass" | "partial" | "fail" | "not_applicable";
export type Severity = "low" | "medium" | "high" | "critical";

export type EtalonEvidence = {
  quote: string;
  location: string;
};

export type EtalonLayer1Item = {
  id: string;
  dimension: string;
  status: CheckStatus;
  severity: Severity;
  title: string;
  summary: string;
  evidence: EtalonEvidence[];
  recommendation: string;
  confidence: number | null;
};

export type EtalonLayer2Item = {
  id: string;
  parent_layer_1_id: string;
  check: string;
  status: CheckStatus;
  severity: Severity;
  finding: string;
  evidence: EtalonEvidence[];
  expected_fix: string;
  confidence: number | null;
};

export type EtalonRecord = {
  id: string;
  document_id: string;
  author_id: string;
  source: "manual" | "ai_post_annotation" | "imported_defense" | "gate2_benchmark";
  source_metadata: Record<string, unknown>;
  document_type: DocumentType;
  real_defense_status: string | null;
  defense_comments: string | null;
  expected_verdict: Verdict;
  layer_1: EtalonLayer1Item[];
  layer_2: EtalonLayer2Item[];
  key_findings: string[];
  forbidden_false_findings: string[];
  status: EtalonStatus;
  version: number;
  raw_file_visible_to_all: boolean;
  created_at: string;
  updated_at: string;
};

export type Gate2BenchmarkImportRequest = {
  benchmark_dir?: string | null;
  activate?: boolean;
};

export type Gate2BenchmarkImportResponse = {
  imported_count: number;
  skipped_count: number;
  parse_enqueued_count: number;
  etalons: EtalonRecord[];
};

export type EtalonUpdatePayload = Partial<{
  expected_verdict: Verdict;
  layer_1: EtalonLayer1Item[];
  layer_2: EtalonLayer2Item[];
  key_findings: string[];
  forbidden_false_findings: string[];
  real_defense_status: string | null;
  defense_comments: string | null;
  raw_file_visible_to_all: boolean;
}>;

export async function listEtalons(): Promise<{ etalons: EtalonRecord[] }> {
  return apiFetch<{ etalons: EtalonRecord[] }>("/etalons");
}

export async function getEtalon(etalonId: string): Promise<EtalonRecord> {
  return apiFetch<EtalonRecord>(`/etalons/${etalonId}`);
}

export async function getAnnotationQueue(): Promise<{ etalons: EtalonRecord[] }> {
  return apiFetch<{ etalons: EtalonRecord[] }>("/annotation/queue");
}

export async function createEtalonDraft(analysisId: string): Promise<EtalonRecord> {
  return apiFetch<EtalonRecord>(`/analyses/${analysisId}/etalon-draft`, { method: "POST" });
}

export async function importPastDefense(form: FormData): Promise<EtalonRecord> {
  return apiFetch<EtalonRecord>("/documents/past-defense", { method: "POST", body: form });
}

export async function importGate2Benchmark(payload: Gate2BenchmarkImportRequest): Promise<Gate2BenchmarkImportResponse> {
  return apiFetch<Gate2BenchmarkImportResponse>("/etalons/import/gate2-benchmark", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function updateEtalon(etalonId: string, payload: EtalonUpdatePayload): Promise<EtalonRecord> {
  return apiFetch<EtalonRecord>(`/etalons/${etalonId}`, { method: "PATCH", body: JSON.stringify(payload) });
}

export async function publishEtalon(etalonId: string): Promise<EtalonRecord> {
  return apiFetch<EtalonRecord>(`/etalons/${etalonId}/publish`, { method: "POST" });
}

export async function archiveEtalon(etalonId: string): Promise<EtalonRecord> {
  return apiFetch<EtalonRecord>(`/etalons/${etalonId}/archive`, { method: "POST" });
}
